### PR TITLE
State machine design

### PR DESCRIPTION
**Pull request by lasshpanzer**
The PR proposes to add statuses for volume, snapshot, and attachment resources.

### CODE REVIEW

**Comment by xing-yang** [comment]
@lasshpanzer Thanks for submitting this PR! It looks like you submitted 4 files.  Please only submit Opensds_state_machine_design.md and remove the other 3 redundant files, especially the word doc should not be submitted here.  Thanks.

File: Opensds_state_machine_design.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+Status: WIP
+
+Version: Alpha
+
+Author(s): jack li (@lasshpanzer)
+
+## Summary
+
+The State machines of volume,snapshot and attachment show the real-time status and state transitions between actions.
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: Add a space between "volume," and "snapshot".

File: Opensds_state_machine_design.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+## Motivation
+
+For example, When creating a volume, there is needed to know if it has been created or if something went wrong.
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: When creating a volume, there is a need to know if it has been created successfully or if something went wrong.

Remove "For example".

File: Opensds_state_machine_design.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+Design volume statuses
+
+Design snapshot statuses
+
+Design attachment statuses
+
+### Non-Goals
+
+Some architectures needs to be changed.
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: This sentence is confusing.  Can you please clarify?

Also we do need statuses for Group, GroupSnapshot, and Replication later.  Please add them here under Non-Goals for this PR and noted that they will be added later.

File: Opensds_state_machine_design.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+## Proposal
+
+There is no mechanism to show the real-time status and state transitions in the current opensds version.It needs to be implemented.
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: Add a space after the period.

File: Opensds_state_machine_design.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+### Data model impact
+
+The state attributes of the data have been added to many new states
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: Status will be added to volume, snapshot, and attachment resources.

File: Opensds_state_machine_design.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+No
+
+### Security impact
+
+Does it affect security?
+No
+
+### Other end user impact
+
+Other than API, are there other changes that affect a user? For example the CLI changes.
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: When getting a resource such as a volume, end user can now see the changed status of the resource.

File: Opensds_state_machine_design.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+Is there any performance impact?
+No
+
+### Other deployer impact
+
+None
+
+### Developer impact
+
+Other developers may use these state attributes to design other properties and methods.
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: Driver developers will need to return status after an operation is completed on a resource.

File: Opensds_state_machine_design.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+None
+
+## Implementation
+
+Volume statuses
+
+Status            Description
+creating	      The volume is being created. 
+available	      The volume is ready to attach to an instance.
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: The volume is ready to be used.

File: Opensds_state_machine_design.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+attaching	      The volume is attaching to an instance.
+detaching	      The volume is detaching from an instance.
+in-use            The volume is attached to an instance.
+deleting	      The volume is being deleted.
+error             A volume creation error occurred.
+error_deleting    A volume deletion error occurred.
+error_extending   An error occurred while attempting to extend a volume.
+extending         The volume is being extended.
+
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: I suggest changing the statuses following Go's naming convention as follows:
creating -> Creating
available -> Available
in-use -> InUse
error_deleting -> ErrorDeleting
......
Similarly make this kind of change to all statuses.

File: Opensds_state_machine_design.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
+# Opensds state machine design
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: OpenSDS

File: Opensds_state_machine_design.md [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+Author(s): jack li (@lasshpanzer)
+
+## Summary
+
+The State machines of volume, snapshot and attachment show the real-time status and state transitions between actions.
+
+## Motivation
+
+When creating a volume, there is needed to know if it has been created or if something went wrong.
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: s/there is needed/there is a need

File: Opensds_state_machine_design.md [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+Status: WIP
+
+Version: Alpha
+
+Author(s): jack li (@lasshpanzer)
+
+## Summary
+
+The State machines of volume, snapshot and attachment show the real-time status and state transitions between actions.
```
Comment (xing-yang) [inline Opensds_state_machine_design.md]: s/show/to show

**Comment by leonwanghui** [comment]
@lasshpanzer IMO there are some confusions in `Proposol` and `Implementation` part, because proposol normally means what's your general plan (architecture level) to solve the question and the implementation part is used to explain the concrete design (code level) to make a clear view of the use of this proposol. Does that make sense?

**Pull request closed without merging**